cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"fmt"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/scheduler/cache"
+	"sync"
+	"time"
+)
+
+// OpenstackNodeGroup implements NodeGroup interface from cluster-autoscaler/cloudprovider.
+//
+// Represents a homogeneous collection of nodes within a cluster,
+// which can be dynamically resized between a minimum and maximum
+// number of nodes.
+type OpenstackNodeGroup struct {
+	openstackManager OpenstackManager
+	id               string
+
+	clusterUpdateMutex *sync.Mutex
+
+	minSize int
+	maxSize int
+	// Stored as a pointer so that when autoscaler copies the nodegroup it can still update the target size
+	targetSize *int
+
+	nodesToDelete      []*apiv1.Node
+	nodesToDeleteMutex sync.Mutex
+
+	waitTimeStep time.Duration
+}
+
+// WaitForClusterStatus checks once per second to see if the cluster has entered a given status.
+// Returns when the status is observed or the timeout is reached.
+func (ng *OpenstackNodeGroup) WaitForClusterStatus(status string, timeout time.Duration) error {
+	klog.V(2).Infof("Waiting for cluster %s status", status)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(ng.waitTimeStep) {
+		clusterStatus, err := ng.openstackManager.GetClusterStatus()
+		if err != nil {
+			return fmt.Errorf("error waiting for %s status: %v", status, err)
+		}
+		if clusterStatus == status {
+			klog.V(0).Infof("Waited for cluster %s status", status)
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout (%v) waiting for %s status", timeout, status)
+}
+
+// IncreaseSize increases the number of nodes by replacing the cluster's node_count.
+//
+// Takes precautions so that the cluster is not modified while in an UPDATE_IN_PROGRESS state.
+// Blocks until the cluster has reached UPDATE_COMPLETE.
+func (ng *OpenstackNodeGroup) IncreaseSize(delta int) error {
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	if delta <= 0 {
+		return fmt.Errorf("size increase must be positive")
+	}
+
+	size, err := ng.openstackManager.NodeGroupSize(ng.id)
+	if err != nil {
+		return fmt.Errorf("could not check current nodegroup size: %v", err)
+	}
+	if size+delta > ng.MaxSize() {
+		return fmt.Errorf("size increase too large, desired:%d max:%d", size+delta, ng.MaxSize())
+	}
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
+	if err != nil {
+		return fmt.Errorf("can not increase node count: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not add nodes, cluster is in %s status", currentStatus)
+	}
+	klog.V(0).Infof("Increasing size by %d, %d->%d", delta, *ng.targetSize, *ng.targetSize+delta)
+	*ng.targetSize += delta
```

Comment:
Can you clarify this for me, wouldn't this operation ignore the lock *ng.targetSize += delta since the lock would only lock the address of the pointer and not the value at the address?
Full disclosure, I'm not the primary on the autoscaler work but why does the value of the targetSize need to be a pointer anyways? i saw the comment about // when autoscaler copies the nodegroup it can still update the target size
When does this actually occur, wouldnt it just ask to use the context ng already has?

Reply:
copies the nodegroup it can still update the target size, when it copies this nodegroup its intended to be just a simple accessor no?

Reply:
I was thrown off by this as well.  The lock is good to ensure state, but the targetSize is being modified for anything outside of the lock?  I think if another method relies on how IncreaseSize modifies the targetSize field, it should just respect the lock.  Sorry if I am just misunderstanding the strategy.

Reply:
What I found was that when targetSize was a non-pointer on the nodegroup struct, it would be updated during the scale up or scale down but when the next loop came around it was as if the value had never been changed. NodeGroupForNode returns a pointer to the nodegroup so maybe it is passed again by the autoscaler in a way that means it updates a copy of targetSize instead of the original.

The clusterUpdateMutex is protecting targetSize, but the main purpose of the lock was for preventing a new scale operation from taking place while another one is still waiting to complete. Since we're discussing (and I'm attempting) changing this to allow new updates while the previous one is still in progress, the lock will need to be released earlier. The targetSize would also have to be updated sooner to allow subsequent updates to add a delta to the correct value, so it would probably still be within the scope of the lock.

Reply:
ng.clusterUpdateMutex.Lock() will only protect the targetSize pointer from being changed not the value it points to so if something outside of your own code (the autoscaler core) changes the value at the pointer thats been exposed it will change regardless of the lock (in other words that copy that was made)
> The clusterUpdateMutex is protecting targetSize, but the main purpose of the lock was for preventing a new scale operation from taking place while another one is still waiting to complete.
 
I definitely agree with the existence and intent of the mutex =) 

We didnt seem to have this issue maybe @scott-chaney can chime in and clarify or give examples

Reply:
My strategy was to lock the cache mutex whenever state was being modified. If any "shared" values need to change, the cache is updated by the function holding the lock.  Anything that needs to read or update any values in the cache respects and maintains the lock. This handles everything within a single refresh loop, and if the state is ever corrupted, it is re-acquired on the next refresh cycle.